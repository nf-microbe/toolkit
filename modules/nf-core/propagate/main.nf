process PROPAGATE {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/vrhyme:1.1.0--pyhdfd78af_1' :
        'biocontainers/vrhyme:1.1.0--pyhdfd78af_1' }"

    input:
    tuple val(meta) , path(fastq)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(coords)

    output:
    tuple val(meta), path("${prefix}.propagate.tsv")    , emit: results
    tuple val(meta), path("${prefix}.propagate.log")    , emit: log     , optional: true
    path  "versions.yml"                                , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    def is_compressed = fasta.getExtension() == "gz" ? true : false
    def fasta_name = is_compressed ? fasta.getBaseName() : fasta
    """
    if [ "${is_compressed}" == "true" ]; then
        gzip -c -d ${fasta} > ${fasta_name}
    fi

    Propagate \\
        -f ${fasta_name} \\
        -v ${coords} \\
        -r ${fastq} \\
        -o ${prefix}.propagate \\
        -t ${task.cpus} \\
        ${args}

    mv ${prefix}.propagate/${prefix}.propagate.tsv ./${prefix}.propagate.tsv
    mv ${prefix}.propagate/${prefix}.propagate.log ./${prefix}.propagate.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        propagate: \$(Propagate --version 2>&1 | sed 's/^.*PropagAtE v//; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.propagate.tsv
    touch ${prefix}.propagate.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        propagate: \$(Propagate --version 2>&1 | sed 's/^.*PropagAtE v//; s/ .*\$//')
    END_VERSIONS
    """
}
