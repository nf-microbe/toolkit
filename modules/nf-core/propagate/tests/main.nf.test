nextflow_process {

    name "Test Process PROPAGATE"
    script "../main.nf"
    process "PROPAGATE"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "propagate"

    test("fastq.gz + fasta.gz + coords.tsv") {
        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ],
                        [
                            file(params.modules_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                            file(params.modules_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("https://github.com/nf-core/test-datasets/raw/refs/heads/phageannotator/modules/local/propagate/b_fragilis_propagate_contigs.fasta", checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("https://github.com/nf-core/test-datasets/raw/refs/heads/phageannotator/modules/local/propagate/b_fragilist_provirus_coords.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.results,
                    process.out.versions).match() }
            )
        }
    }

    test("fastq.gz + fasta.gz + coords.tsv - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ],
                        [
                            file(params.modules_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                            file(params.modules_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("https://github.com/nf-core/test-datasets/raw/refs/heads/phageannotator/modules/local/propagate/b_fragilis_propagate_contigs.fasta", checkIfExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("https://github.com/nf-core/test-datasets/raw/refs/heads/phageannotator/modules/local/propagate/b_fragilist_provirus_coords.tsv", checkIfExists: true)
                    ]
                )
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
