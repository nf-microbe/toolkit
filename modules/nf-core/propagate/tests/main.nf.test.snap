{
    "fastq.gz + fasta.gz + coords.tsv - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.propagate.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.propagate.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,67ff2549247db9f84c3b0bb539876381"
                ],
                "log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.propagate.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "results": [
                    [
                        {
                            "id": "test"
                        },
                        "test.propagate.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,67ff2549247db9f84c3b0bb539876381"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-15T12:59:13.594299418"
    },
    "fastq.gz + fasta.gz + coords.tsv": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.propagate.tsv:md5,546afd777a6782677163130a5efb2869"
                ]
            ],
            [
                "versions.yml:md5,67ff2549247db9f84c3b0bb539876381"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-15T13:24:14.614754547"
    }
}